### ISSUE TITLE
Huge Lag/Rotations Not Working

### ISSUE THREAD

**Issue by Fot3k** [Romain-P/Ewt-Scripts#4]
Hi,

Been messing around trying to get this to work correctly for my Retribution Paladin on 3.3.5a Private Server. When I enable it in game and go into WSG or anything global really it becomes impossible to play due to lag. I know you mention there is zero lag but this is not the case for me. I can't tell what it is doing because of this.

I have a high end system: Ryzen 1800X / 16GB 3200GHz / SSD etc.

I would love to use the rotations you have implemented but they don't seem to be working. I am using PQR at the moment but doesn't quite feel as best as it could be. Your script from what I can read looks to be much more polished and what I am after.

Could you offer some guidance please on how to get this working.

Much appreciated!

**Comment by Romain-P** [comment]
Hi!
This script can't be executed from a standard PQR profile. You need something like a wrapper.. 

Anyway, you shouldn't use PQR. Why don't you use EWT and just follow my readme? :")
You're free to send me an email with your discord/skype, I will help you 

Cheers

**Comment by Fot3k** [comment]
Hi, 

I did follow your readme. I created a macro in game to execute the command and says it's activated but it still does nothing but cause lag. You mention nothing else than to put the script folder in my wow directory and execute the script.

I've been through each file and read the config.

I only use PQR because this doesn't work for me right now and it seems impossible to find any other EWT scripts for 3.3.5 pally.

Thanks.

**Comment by Romain-P** [comment]
Hi!

It is not working due of the fact that my scripts are using the ewt's API.
EWT provides a lot of useful functions such as ray-casting (for LOS checks), map objects etc..

When such a function is used, the lua wrapper meets and throws an error, because the function is unknown.

You have to use EWT to enjoy these functions, and therefore my scripts...

Cheers,
Don't hesitate to share your discord or skype ;)

**Comment by Fot3k** [comment]
I am using EWT and the script runs in game with EWT but like I said it doesn't do anything but cause lag. I know how to use EWT, I have LUA unlocked with everything mode.

I've tested it on single NPC's just stand there and also in a duel and WSG battleground.

**Comment by Romain-P** [comment]
Alright I thought you were using PQR,
I will test it as soon as possible, it can be due of an EWT release.. I'll look for deprecated methods or idk,

Give me the week ;)
Cheers

**Comment by Romain-P** [comment]
Try with the current master?

**Comment by Fot3k** [comment]
Lag seems to be fixed, only tested in world at the moment though until tomorrow for battlegrounds.

Still not doing anything at all though with abilities etc, no rotations? Unless you've not worked on that yet.

Thanks for the update :)

**Comment by Romain-P** [comment]
Ye I didn't released the ret profile officially, but it's working like the priest one

You can access multiple functions such as 
 * Dps rotation : https://github.com/Romain-P/Ewt-Scripts/blob/master/paladin_ret.lua#L232
 * Safe cast : https://github.com/Romain-P/Ewt-Scripts/blob/master/paladin_ret.lua#L249
 * Judgement function: https://github.com/Romain-P/Ewt-Scripts/blob/master/paladin_ret.lua#L205

> Dps rotation

To use the dps rotation, create a macro with `/script Dps(target)` , then you can press it again and again targetting an enemy. Yes, it's semi-auto, when you want to stop the rotation for a custom action, just stop spamming the macro and do your business ;)
It should perform the best rotation for damaging.   
  

> Safe Cast

The safe cast function is a spell-cast-wrapper that allows you to avoid to use your important spells on protected enemies.
For example, make a macro like `/script SafeCast(PaladinSpells.HOJ, target)`. It will cast Hammer of Justice on your target when the macro iis pressed unless if the target is protected e.g by Divine Shield, Cyclone, Grounding Totem on, Reflect etc...

> Judgement

This is a simple feature, make a macro like` /script Judgement(target)` and when pressed, it will use Judgement of Justice on rogues/druids, Judgement of Light otherwise.

> Common

You can find all spells into https://github.com/Romain-P/Ewt-Scripts/blob/master/shared_spells.lua  
Don't hesitate to fill it if some spells are missing. Use it as much as possible to make the code clear and human-readable ;)

In the same extent, use defined units as much as possible to save memory and increase the celerity of the script.
Everything defined is accessible by your macros.

> About automatic features

You can enable or disable any automatic features in the configuration at the begin of the file.  
For example, you can find the configuration for auto-friendly-dispel: https://github.com/Romain-P/Ewt-Scripts/blob/master/paladin_ret.lua#L95
Change` ENABLED = true `by `ENABLED = false` for disable it.

> Ret specific features
 * https://github.com/Romain-P/Ewt-Scripts/blob/master/paladin_ret.lua#L176 This feature kills automatically the grounding totem with Taunt or your Weapon (if in range). It's full auto, when a grounding totem is ON on your target, the script will destroy the grounding totem if you have enough cds for this. You can disable it writing false.

 * https://github.com/Romain-P/Ewt-Scripts/blob/master/paladin_ret.lua#L178 You also have auto taunt pet, that uses Taunt on any pet in range in an arena game. Free to you to disable it too.

**Comment by Romain-P** [comment]
PS: don't hesitate to request new  features by creating an issue, tell me if I can close this one ;)

Cheers

**Comment by Fot3k** [comment]
Thanks for the detailed instructions much appreciated, I got the DPS rotation working with the macro. :)

Few things I've noticed:

- It uses Divine Storm when I am no where near the target and is the first spell it uses. Would obviously be better if it used a Judgement first unless in range.

- Doesn't seem to auto target when in range?

Is there anyway to make the rotation and the above happen automatically? That's why I like PQR, I can just run in and it auto targets, buffs and does the rotation with a Hammer of Wrath to finish. Also snipes players I'm not focusing with Hammer of Wrath. I always get top of the leaderboard in WSG with this. I just feel the defence side is not so good, it casts sacred shield but doesn't seem to cleanse too well or use Divine Shield and Lay of Hands etc. Trinket would be amazing feature if possible for PVP when movement impaired.

Basically I use the moving cast feature in EWT with PQR which frees up my fingers to move around instead so I don't have to be mashing buttons for abilities while stood still. Hope that makes sense? Gives me a big advantage, good for kiting!

I can see and understand that your scripts are much cleaner and thought out but I just need the auto features to replace PQR. I would also be happy to donate if we can get this working. :)

Thanks.
